core/test/config/config.cpp

```diff
@@ -0,0 +1,189 @@
+    reg.emplace("precond", this->mtx);
+
+    pnode p{{{"generated_preconditioner", pnode{"precond"}},
+             {"criteria", this->stop_config}}};
+    auto obj =
+        parse<LinOpFactoryType::Cg>(p, reg, type_descriptor{"float32", "void"})
+            .on(this->exec);
+
+    ASSERT_NE(dynamic_cast<gko::solver::Cg<float>::Factory*>(obj.get()),
+              nullptr);
+    ASSERT_NE(dynamic_cast<gko::solver::Cg<float>::Factory*>(obj.get())
+                  ->get_parameters()
+                  .generated_preconditioner,
+              nullptr);
+}
+
+
+TEST_F(Config, GenerateObjectWithPreconditioner)
+{
+    auto reg = registry();
+    auto precond_node =
+        pnode{{{"type", pnode{"solver::Cg"}}, {"criteria", this->stop_config}}};
+    pnode p{{{"value_type", pnode{"float64"}},
+             {"criteria", this->stop_config},
+             {"preconditioner", precond_node}}};
+
+    auto obj = parse<LinOpFactoryType::Cg>(p, reg).on(this->exec);
+
+    ASSERT_NE(dynamic_cast<gko::solver::Cg<double>::Factory*>(obj.get()),
+              nullptr);
+    ASSERT_NE(dynamic_cast<gko::solver::Cg<double>::Factory*>(obj.get())
+                  ->get_parameters()
+                  .preconditioner,
+              nullptr);
+}
+
+
+TEST_F(Config, GenerateObjectWithCustomBuild)
+{
+    configuration_map config_map;
+    config_map["Custom"] = [](const pnode& config, const registry& context,
+                              const type_descriptor& td_for_child) {
+        return gko::solver::Bicg<double>::build().with_criteria(
+            gko::stop::Iteration::build().with_max_iters(2u));
+    };
+    auto reg = registry(config_map);
+    auto precond_node =
+        pnode{std::map<std::string, pnode>{{"type", pnode{"Custom"}}}};
+    pnode p{{{"value_type", pnode{"float64"}},
+             {"criteria", this->stop_config},
+             {"preconditioner", precond_node}}};
+
+    auto obj =
+        parse<LinOpFactoryType::Cg>(p, reg, type_descriptor{"float64", "void"})
+            .on(this->exec);
+
+    ASSERT_NE(dynamic_cast<gko::solver::Cg<double>::Factory*>(obj.get()),
+              nullptr);
+    ASSERT_NE(dynamic_cast<const gko::solver::Bicg<double>::Factory*>(
+                  dynamic_cast<gko::solver::Cg<double>::Factory*>(obj.get())
+                      ->get_parameters()
+                      .preconditioner.get()),
+              nullptr);
+}
+
+
+TEST(GetValue, IndexType)
+{
+    long long int value = 123;
+    pnode config{value};
+
+    ASSERT_EQ(get_value<int>(config), value);
+    ASSERT_EQ(get_value<long>(config), value);
+    ASSERT_EQ(get_value<unsigned>(config), value);
+    ASSERT_EQ(get_value<long long int>(config), value);
+    ASSERT_EQ(typeid(get_value<int>(config)), typeid(int));
+    ASSERT_EQ(typeid(get_value<long>(config)), typeid(long));
+    ASSERT_EQ(typeid(get_value<unsigned>(config)), typeid(unsigned));
+    ASSERT_EQ(typeid(get_value<long long int>(config)), typeid(long long int));
```

Comment:
[Type assertions](https://google.github.io/googletest/advanced.html#type-assertions) do exist in Google test and they are at compile-time, but show a pass during runtime, so I would prefer if you use these:
```suggestion
    testing::StaticAssertTypeEq<decltype(get_value<int>(config)),int>();
    testing::StaticAssertTypeEq<decltype(get_value<long>(config)), long>();
    testing::StaticAssertTypeEq<decltype(get_value<unsigned>(config)), unsigned>();
    testing::StaticAssertTypeEq<decltype(get_value<long long int>(config)), long long int>();
```